### ISSUE TITLE
Contact groups can't be deleted

### ISSUE THREAD

**Issue by DMTGMBH** [nextcloud/contacts#1175]
There's is no way to delete a contact group after it was created. Create a contact, add a new group. Remove the group after saving from the contact. Now the group is listed in the left sidepanel and there's no way to delete it.

**Comment by pwFoo** [comment]
Can't delete it manually, but if there is no contact in the group it will be removed automatically.
Seems to work as should?

**Comment by skjnldsv** [comment]
Hi!
Until we get #213, adding such thing will trigger one request per contact within the group as the current way is really not ok (performance speaking)

I don't want to add a functionnality that will break the user experience :/

So until then, we'll have to wait! :)

**Comment by skjnldsv** [comment]
Re opening for search purpose

**Comment by ronnicek** [comment]
Having the same issue. Created group with typo.. and then group is empty and there is no possibility to delete it :)

**Comment by skjnldsv** [comment]
@ronnicek you can still remove it on the contacts that have it :)
Just uncheck it from the groups dropdown and write a new one

**Comment by ronnicek** [comment]
I am able to do that of course, but then the group with typo is still there in the list of groups :)

**Comment by skjnldsv** [comment]
@ronnicek it means at least one contact still uses the old group :)

**Comment by ronnicek** [comment]
@skjnldsv 🤦‍♂️. Really. Sorry then, solved for me :)

**Comment by bendem** [comment]
When I imported my contacts from google, they all had the group myContacts, I'd really like to be able to delete that group without having to go through each contact one by one.

**Comment by droumanet** [comment]
It's a pain to have an unused group displayed in the left view... so imagine when you've more unused groups and have to scroll down to find the latest good one.
Please, add a way to delete them 😉 :

**Comment by MoritzHorch** [comment]
@bendem 
What worked for me is to export as .vcf from Google and with any text editor of your choice replace the corresponding tag in the file with a blank line. Wish it was easier, but works until this gets implemented.

**Comment by skjnldsv** [comment]
After #1687 it might be easier to implement such feature :)

**Comment by tvollscw** [comment]
> @skjnldsv man_facepalming. Really. Sorry then, solved for me (after refresh) :)

Had to come here to see how this was done. Doesn't seem very intuitive. How about add a trash can icon beside the group if it is created, and then abandoned? In my case, I named the group incorrectly and wanted to rename, but then thought I couldn't rename or delete the group! 

It is a small change, but feels like it makes more sense to end users.

**Comment by skjnldsv** [comment]
> Doesn't seem very intuitive. How about add a trash can icon beside the group if it is created, and then abandoned? In my case, I named the group incorrectly and wanted to rename, but then thought I couldn't rename or delete the group!

100% agree on :+1: 
Like I said, any kind of batch modification of contacts (groups) will be easier after #1687 :wink: 

> After #1687 it might be easier to implement such feature :)

**Comment by krafel** [comment]
> @skjnldsv man_facepalming. Really. Sorry then, solved for me (after refresh) :)

Hmmm... for me, I cannot uncheck the group from the last member of it. I mean: I can, but after refresh, it's still member! So the group remains at size 1 an cannot be deleted. Buggy.

**Comment by jivanpal** [comment]
@krafel said:

> after refresh, it's still member! So the group remains at size 1 an cannot be deleted. Buggy.

Same issue for me with Nextcloud 20.0.0, Contacts 3.4.0.

Suppose the contact "Alice" is the only contacts in the groups "Friends" and "Colleagues", then Alice can be successfully removed from "Friends" only if the removal is performed whilst viewing Alice in the "Colleagues" or "All contacts" sections, but not when viewing Alice in the "Friends" section. It seems that if we remove Alice from "Friends" whilst viewing her in "Friends", then that act of the contact disappearing from the list we're viewing somehow interferes with the intended act of removing her from the group.

**Comment by ldmpub** [comment]
+1 (I'm concerned by this anti-functionnality)

**Comment by dargelm** [comment]
Unbelievable that such a simple functionality is missing and no progress is made for such a long time.

**Comment by the-rocinante** [comment]
Looks like #1687 was merged 6 months ago. Any word on this?

Fyi, I just created a group in contacts which was misspelled, and can neither rename nor remove it in spite of it being empty. I'm tempted to try and remove it from the database manually, but past efforts to manually modify app data have been considerably less than encouraging.

**Comment by skjnldsv** [comment]
> Looks like #1687 was merged 6 months ago. Any word on this?

No time for this unfortunately. Though the functionnality is far more easier to add now.
Anyone willing to dive into this, I can point and assist a bit! :rocket:

**Comment by JezzaHehn** [comment]
> Anyone willing to dive into this, I can point and assist a bit!

Can you point us in the right direction? :green_heart:

**Comment by skjnldsv** [comment]
Of course :)

Each contact needs to be edited for this.
There is already a similar process when we batch-add contacts to a group.
Doing the exact opposite would delete the group: 
https://github.com/nextcloud/contacts/blob/master/src/views/Processing/AddToGroupView.vue

Magic happens here:
https://github.com/nextcloud/contacts/blob/c23a7fdf36d52af269b0593cbaf363fe860c0245/src/views/Contacts.vue#L761-L810

The selection here would be all the contacts of the group (you can get them with `group.contact` and then map the contacts keys to `this.contacts` 
![image](https://user-images.githubusercontent.com/14975046/116533167-57dad800-a8e1-11eb-92b8-2b862f754328.png)

https://github.com/nextcloud/contacts/blob/f44028131344636e45c5158cc13ccbe4edd19097/src/services/appendContactToGroup.js#L29-L39

After that there is already a vuex action to remove a contact from a group https://github.com/nextcloud/contacts/blob/d8ede82510afe55a06c754710463a45eab6e5d7c/src/store/groups.js#L151-L161

**Comment by frank42hh** [comment]
I can't believe that this is still an open thing after almost two years ... I hope this is getting fixed soon since it's quite a pain when you don't know this and import a vcf from a google contacts export and end in having dozens of unwanted contact groups while not being able to remove them.

**Comment by ferdieb** [comment]
> There's is no way to delete a contact group after it was created. Create a contact, add a new group. Remove the group after saving from the contact. Now the group is listed in the left sidepanel and there's no way to delete it.

How about simply a function to perhaps not delete the group but select another group from a dropdown of existing other groups and then (choose to) replace or add all the current group members to the selected group? This should be an update query for the related group members - even if it happens similar to theming being applied to all users.

**Comment by elbubar** [comment]
Hi guys. Thanks for the job done. Nextcloud is really something.

What would be the best way to brute-delete the address book so that people could at least start from a fresh install of Contacts and reimport a trimmed .vcf or .CSV ?

This, linked with a very simple Msgbox (sorry I know mainly VBA excel...) Alerting people on group creation as "Beware in choosing the group name, it still can't be deleted or change,  but this functionality will come one day !"

No offense here. Knowing it will prevent people getting frustrated and having the opportunity to start from scratch in Contacts and trim the reimport will make already frustrated ones a bit grumpy for an afternoon and go to sleep happy.

**Comment by deefrawley** [comment]
Hi All. This issue has now been open for 30 months - where is it at? When can this be resolved as it currently is a poor user experience.

**Comment by frank42hh** [comment]
I don't think that this ever gets fixed since it looks like nobody cares anymore about "user experience". I'm just surprised that there's still no "Remember this is free software, so take it or leave it" comment so far.

**Comment by mclang** [comment]
I accidentally mistyped group name, and now I found out that year is 2022 and contact groups cannot be deleted :open_mouth: 
Well I hope this improvement is int **this year's** road map :pray:

**Comment by miaulalala** [comment]
Hi peeps.

> nobody cares anymore about "user experience"

We do. But just as everywhere else, there's only 24 hours in a day, and there's only so many people on the team.

> I'm just surprised that there's still no "Remember this is free software, so take it or leave it" comment so far.

I know these comments are annoying to people, but it's the truth. I would absolutely love it if I had the time or spoons to take care of every issue that crops up, but I don't, and neither do my fellow devs.

So I would ask you all to please keep in mind that this isn't personal, or to be taken as an offence if we don't work on things.

If any of you would like to help, feel free to give this a shot and I'll make sure your PR will be merged as soon as possible.

**Comment by frank42hh** [comment]
> 

You're right, I'm sorry. It's just... it's very frustrating when you're trying to use a specific software and then happen to have such a massive showstopper, which breaks your entire migration project. With this contact groups problem in mind, I can't use the contacts module at all and without the contacts module the entire switch to nextcloud doesn't make any sense for me now.

**Comment by theneilkirk** [comment]
I thought I'd add that it's not that Groups cannot be deleted. If you remove the Group from all the Contacts in that Group, then it ceases to exist (I believe). But yes, you need to do that for each Contact. Totally get this is an issue if you have lots of Contacts in a Group.

But while it may not be a great user experience it is not, for me, enough of a problem to mean I don't get to keep my contacts off of commercial data-stealing platforms.

**Comment by jsiei97** [comment]
I found a workaround for this problem (where it's not possible to remove empty groups).
 
I have my contacts synced with thunderbird, and this software actually has a option to delete a group, so I tried that, and synced the data. And then the empty group was gone in the nextcloud web interface.

**Comment by ldmpub** [comment]
Nice find @jsiei97 👍 (I'm also running Thunderbird with NextCloud, I will use this workaround for now on!).

**Comment by heegard** [comment]
> I found a workaround for this problem (where it's not possible to remove empty groups).
> 
> I have my contacts synced with thunderbird, and this software actually has a option to delete a group, so I tried that, and synced the data. And then the empty group was gone in the nextcloud web interface.

Great minds think alike.  In fact, I found the thunderbird editor to be more useful in general!

**Comment by MikulasVanousek** [comment]
For some reason, it is impossible for me to remove the last person from a group. It appears to work, but when I then refresh the page, the last user is back in the group.

**Comment by dargelm** [comment]
I know this behavior. I tried it some times. Anytime and some browser refreshes or logouts / logins later it works.


-- 
Marko Dargel
Balkhauser Weg 57A
45529 Hattingen

📱: +49 160 96879574‬
📞: ‭+49 2324 90227579
🌍 : https://myancestry.de 
📧: ***@***.***

**Comment by BOFH90** [comment]
Not a fix, but a workaround. You could delete it with SQL. They are in "oc_addressbooks".

**Comment by Rik44444** [comment]
has this been addressed yet? I also cannot seem to delete groups

**Comment by Jerome-Herbinet** [comment]
I confirm the presence of this bug which is quite annoying. 
Nextcloud 23.0.4
Contacts 4.1.0

**Comment by rcheetah** [comment]
> I don't want to add a functionnality that will break the user experience :/

i think not being able to edit group is an even worse user experience :( 
I have a lot of strange auto-generated groups (i don't know if they were added by nextcloud or not, but I never saw them before), and I'm not able to remove the groups.

**Comment by Jerome-Herbinet** [comment]
"Recently contated" should be the only one that can't be deleted.

**Comment by Jerome-Herbinet** [comment]
> > I don't want to add a functionnality that will break the user experience :/
> 
> i think not being able to edit group is an even worse user experience :( 
> I have a lot of strange auto-generated groups (i don't know if they were added by nextcloud or not, but I never saw them before), and I'm not able to remove the groups. 

"Recently contated" should be the only one that can't be deleted.

**Comment by rcheetah** [comment]
I fond out a group gets removed when it's empty. It's just so, that you can't remove the last contact in a group directly. 
When you're in a group and remove the "group" tag manually, the the view immediately updates, because the group would be empty, and it gets removed from the sidebar. But this also closes the main view. Contact data gets saved asynchronously after editing a field. But as the detail view immediately closes, it doesn't really save the data. Therefore at first glance the group is removed, but if you reload it is still there (because the data was never saved to the contact item). 

The solution is to open the last contact directly (via "all contacts" or search bar) and remove the group tag from there. This way the view does'nt get closed immediately and the data will be saved. Then if no contact is left with this group tag, the group will disappear. 

This is not a good solution in any way. You still have to remove the group tag from every single contact (which is'nt very fun for a group with hundreds of items), plus you have to do a workaround to delete the last item.

**Comment by VibrantTreeStump** [comment]
Another workaround for people who want to import large amounts of contacts and want to strip them of their groups beforehand is to open the .vcf file with a program like Notepad++ and delete all lines containing a category with a regular expression.

The regular expression for that job could look like this: ^CATEGORIES:.*\r\n

![Unbenannt](https://user-images.githubusercontent.com/82000085/194729146-91463ad4-ab75-4a1d-a79a-f50ca81d1c6a.PNG)

Leave the field for "Replace with" empty and replace all. Now you're good to go.

**Comment by jivanpal** [comment]
I just want to point out that the issue that @krafel and I had, which I described in [this comment](https://github.com/nextcloud/contacts/issues/1175#issuecomment-707415051), was tracked as #1813 and is now resolved.

It would still be nice to have this feature/enhancement to delete a group in one action.

**Comment by AkechiShiro** [comment]
Hi @skjnldsv,

I'd like to take a stab at working on this, I'm not much experienced with Vue.js and working on Nextcloud apps or Nextcloud itself, but it looks like it's following the MVC model I guess, but in order to start experimenting, do you have any recommendation about setting up the dev environment ? 

Let's say I have a Nextcloud instance running inside a VM [(as well as following probably this guide)](https://docs.nextcloud.com/server/latest/developer_manual/getting_started/devenv.html).

1. Do I just clone the Nextcloud's latest GitHub source or would the latest stable be fine for this feature ?

2. Install the contact app through the administrative panel ? Or should I, git clone the application inside Nextcloud ? (best to work on a custom branch would be to install a stable Nextcloud and then git clone the application I'd guess ?) 

In order for this feature to work, a new service file such as `deleteContactGroup.js` need to be created, by using as a starting point : https://github.com/nextcloud/contacts/blob/f44028131344636e45c5158cc13ccbe4edd19097/src/services/appendContactToGroup.js#L29-L39

I don't understand what is `axios` ? And why is it used to make some kind of request, the beginning of the code makes sense for what it is doing.

A new function needs to be added in `contacts.vue` or should this function handle all actions related to batch deleting and batch adding to groups contacts ?? https://github.com/nextcloud/contacts/blob/c23a7fdf36d52af269b0593cbaf363fe860c0245/src/views/Contacts.vue#L761-L810

I believe this would look like something like this ? 
I'm not sure if in `onContactPickerPick(selection)` both actions can be handled bulk `add` and bulk `remove` to a group

```js
 onContactPickerPick(selection) { 
 	console.debug('Remove', selection, 'to group', this.contactPickerforGroup) 
 	const groupName = this.contactPickerforGroup.name 

 	...

 	// create the array of requests to send 
 	selection.map(async entity => { 
 		try { 
 		    // Get all contacts here somehow
 		    const contact = this.contacts[entity.id] 

                  // loop to requests.push until we have no contact here to treat ??
                  // push contact to server and use limit 
 		  requests.push(limit(() => deleteContactOfGroup(contact, groupName) // service file that needs to be created.
 				.then((response) => { 
                                         // uses the already existing remove contact vuex action?
 					this.$store.dispatch('removeContactToGroup', { groupName, contact }) 
 					this.processStatus.progress++ 
 					this.processStatus.success++ 
 				}) 
 				.catch((error) => { 
 					this.processStatus.progress++ 
 					this.processStatus.error++ 
 					console.error(error) 
 				}) 
 			)) 
 		} catch (e) { 
 			console.error(e) 
 		} 
 	}) 
  
 	Promise.all(requests).then(() => { 
 		this.isProcessDone = true 
 		this.showContactPicker = false 

 		// Auto close after 3 seconds if no errors 
 		if (this.processStatus.failed === 0) { 
 			setTimeout(this.closeProcess, 3000) 
 		} 
 	}) 
 }, 
```

**Comment by dargelm** [comment]
Hi Anna,

 >if anyone knows of any other solution in order to quickly setup a 
Nextcloud dev environment, feel free to let me know.

...try Nethserver. One of the installable applications is Nextcloud.
Setting up a new server takes 30' minutes, a working Nextcloud instance 
takes 3 minutes.
https://www.nethserver.org/learn-more/

The community itself and especially the participation of the developers 
is light years better than Nextcloud.

Release 8 is planned for next year, on a completely new architecture.
https://www.nethserver.org/nethserver-8/

Nethserver runs perfectly in Proxmox.

Sincerely, Marko


------ Originalnachricht ------
Von "Anna" ***@***.***>
An "nextcloud/contacts" ***@***.***>
Cc "MaDa" ***@***.***>; "Comment" 
***@***.***>
Datum 31.08.23, 15:45:24
Betreff Re: [nextcloud/contacts] Contact groups can't be deleted (#1175)

**Comment by Matteo85** [comment]
> Hi! Until we get #213, adding such thing will trigger one request per contact within the group as the current way is really not ok (performance speaking)
> 
> I don't want to add a functionnality that will break the user experience :/
> 
> So until then, we'll have to wait! :)

It sounds like Apple isn't following the vCard 3.x or 4.x standards. Since Nextcloud is built to sustain those standards, waiting to fix an Apple mistake doesn't serve the community, especially since simple issues like being able to delete a user group after creating it (a simple CRUD fix) is being held back by an issue that may never be resolved. Apple has a tendency to do things the way they want and try to force compliance with their modus operandum. Since it is Nextcloud's intention to follow the vCard standards:

With reference to this issue, NextCloud admins said this:  _This is a complicated answer. We did not invent the way contacts works. We are following the official vCard format. This format include a predefined set of properties that other applications/devices supports too (Android, iOS, Windows, Gnome...). Adding a custom property would just make it compatible with Nextcloud and will not be understood by any other clients and make it complicated for us to ensure its sustainability over time. Long story short, we suggest you use the Notes field to add your custom data :)_

I would suggest moving forward with vCard standards UI and functionality issues and not wait for Apple. You can't please everyone, and this is a glaring error (https://github.com/nextcloud/contacts/issues/1175)

**Comment by jivanpal** [comment]
Unless the data for each contact is being stored in vCard format internally (i.e. in the database), I see no reason why #213 is a blocker. Surely the internal format is a bespoke SQL schema, with vCard-formatted data merely generated on the fly if/when an export is requested? If not, that seems like a massive, fundamental oversight in the application design.

**Comment by joshtrichards** [comment]
> Unless the data for each contact is being stored in vCard format internally (i.e. in the database), I see no reason why https://github.com/nextcloud/contacts/issues/213 is a blocker. Surely the internal format is a bespoke SQL schema, with vCard-formatted data merely generated on the fly if/when an export is requested? If not, that seems like a massive, fundamental oversight in the application design.

@jivanpal There's no need to speculate. The code is right in front of you. :-)

**Comment by jivanpal** [comment]
If I had time to read the code, I'd also likely have the time to submit a PR 😊

**Comment by nano13** [comment]
If you are not able to solve this problem in a useful way, I suggest as a workaround to simply remove the functionality of contact groups altogether. That would be a thousand times better than this junk. 

It would also be much easier to explain the absence of this feature to users (differences in use/implementation between different manufacturers, omitting it as the lowest common denominator) than the fact that it is possible to create but not rename or delete contact groups. Because the status quo is unfortunately ridiculous.

**Comment by davidroehrig** [comment]
I am glad to see it has been moved. is there realy not even a workaround to ged rid of groups?  I am looking forward for this being solved (i hope renaming groups also becomes a thing!)
Thanks in advance to the devs!

**Comment by ChristophWurst** [comment]
Delivered with https://github.com/nextcloud/contacts/pull/3957.
